py4j-java/src/main/java/py4j/Protocol.java

```diff
@@ -449,16 +449,17 @@ public final static Object getReference(String commandPart, Gateway gateway) {
 		return gateway.getObject(reference);
 	}
 
-	public final static Object getReturnValue(String returnMessage, Gateway gateway) {
-		Object returnValue = null;
-
+	public final static Object getReturnValue(String returnMessage, Gateway gateway) throws Throwable {
+		final Object result = getObject(returnMessage.substring(1), gateway);
 		if (isError(returnMessage)) {
-			throw new Py4JException("An exception was raised by the Python Proxy. Return Message: " + returnMessage);
+			if (result instanceof Throwable) {
+				throw (Throwable) result;
```

Comment:
I think I would prefer to chain the exception: `throw new Py4JException("An exception...", result)` so the caller knows that something involving a Python proxy went wrong. Even if the exception originated from the Java code, it still went through a Python proxy.

Reply:
Yeah, thinking more about it, `throw (Throwable) result` would break backward compatibility because user code currently only expects Py4JException and not any other throwable.

Reply:
Fair enough. The user can always do this unwrapping themselves if they wish to.

Reply:
Thanks for the explanation, I wanted to do something similar at one point but got lost in a more convoluted solution. Yours is definitively simpler and more elegant.

Backward compatibility is extremely important for Py4J users (we could literally break more than 100k user programs) so could you:

1. add a parameter in CallbackServerParameters/PythonParameters: propagateJavaException=False (by default we keep the former behavior and only add the error message).
2. before sending back the java exception, check the value of the parameter.
3. add yourself to the authors text file :-)

The parameter will eventually be True by default (for 0.11 or 1.0 release).

Thanks!

Reply:
Sure thing: I've added this flag. It does make sense to provide an escape hatch people can use to get back the old behaviour if they really need it.

Thanks for your feedback!